### PR TITLE
Improve Variable guide

### PR DESCRIPTION
**Pull request by kmaehashi**
* Add "Higher-Order Derivatives" section to show usage of `Variable.grad_var`
* Mention `chainer.grad`
* Use `Variable.array` rather than `Variable.data`
* Fix dead links

### CODE REVIEW

**Comment by hvy** [comment]
jenkins, test this please.

File: docs/source/guides/variables.rst [hunk 1]
```diff
@@ -4,28 +4,34 @@ Variables and Derivatives
...
 .. doctest::
 
-   >>> y.data
+   >>> y.array
    array([16.], dtype=float32)
 
+.. note::
+
+   :attr:`~chainer.Variable` has two attributes to represent the underlying array: :attr:`~chainer.Variable.array` and :attr:`~chainer.Variable.data`.
+   They are no difference between two; both are exactly the same object.
```
Comment (hvy) [inline docs/source/guides/variables.rst]: I think it should be `between the two;`.

File: docs/source/guides/variables.rst [hunk 2]
```diff
@@ -4,28 +4,34 @@ Variables and Derivatives
...
+   However it is suggested not to use ``.data`` to avoid confusing :class:`~chainer.Variable` with :class:`~numpy.ndarray`, which has ``.data`` attribute.
```
Comment (hvy) [inline docs/source/guides/variables.rst]: `which also has a` is probably slightly clearer.

File: docs/source/guides/variables.rst [hunk 3]
```diff
@@ -35,7 +41,7 @@ This is done by calling its :meth:`~Variable.backward` method:
    >>> y.backward()
 
 This runs *error backpropagation* (a.k.a. *backprop* or *reverse-mode automatic differentiation*).
-Then, the gradient is computed and stored in the :attr:`~Variable.grad` attribute of the input variable ``x``:
+Then, the gradient is computed and stored in the :attr:`~chainer.Variable.grad` attribute of the input variable ``x``:
```
Comment (hvy) [inline docs/source/guides/variables.rst]: Should we mention `grad_var`?

Comment (kmaehashi) [reply docs/source/guides/variables.rst]: Thanks, that makes sense. I added `Higher-Order Derivatives` section to show the usage of `grad_var`.

**Comment by kmaehashi** [comment]
@niboshi, @hvy Thanks for review, I fixed accordingly.

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
Jenkins, test this please.

File: docs/source/guides/variables.rst [hunk 4]
```diff
@@ -70,6 +76,46 @@ This is done simply by setting the :attr:`~Variable.grad` attribute of the outpu
...
 
+.. note::
+
+   Instead of using :func:`~chainer.Variable.backward`, you can also calculate gradients of any variables in a computational graph w.r.t. any other variables in the graph using the :func:`chainer.grad` function.
+
+
+Higher-Order Derivatives
+------------------------
+
+:class:`~chainer.Variable` also supports higher-order derivatives (a.k.a. double back-propagation).
```
Comment (hvy) [inline docs/source/guides/variables.rst]: Let's stick to `backpropagation` without the hyphen!

File: docs/source/guides/variables.rst (hunk 3 above)
Comment (hvy) [reply docs/source/guides/variables.rst]: Great! This was never documented before I guess.

File: docs/source/guides/variables.rst [hunk 5]
```diff
@@ -70,6 +76,46 @@ This is done simply by setting the :attr:`~Variable.grad` attribute of the outpu
...
+    >>> y.grad = np.ones((2, 3), dtype=np.float32)
+    >>> y.backward(enable_double_backprop=True)
+    >>> x.grad_var
+    variable([[  0.,  12.,  27.],
+              [ 48.,  75., 108.]])
+    >>> assert x.grad_var.array is x.grad
+    >>> assert (x.grad == (3 * x**2).array).all()
+
+:attr:`chainer.Variable.grad_var` is a :class:`~chainer.Variable` for :attr:`chainer.Variable.grad` (which is an :class:`~numpy.ndarray`).
+By specifying ``enable_double_backprop=True`` to ``backward()``, a computational graph for the backward calculation is recorded.
```
Comment (hvy) [inline docs/source/guides/variables.rst]: `By passing` sound more natural (than `specifying ... to ...`). What do you think?

File: docs/source/guides/variables.rst [hunk 6]
```diff
@@ -70,6 +76,46 @@ This is done simply by setting the :attr:`~Variable.grad` attribute of the outpu
...
+
+
+Higher-Order Derivatives
+------------------------
+
+:class:`~chainer.Variable` also supports higher-order derivatives (a.k.a. double back-propagation).
+
+Let's see a simple example.
+First calculate the first-order derivative.
+Note that ``enable_double_backprop=True`` is specified to ``y.backward()``.
```
Comment (hvy) [inline docs/source/guides/variables.rst]: Ditto, `passed to`.

**Comment by hvy** [comment]
Thanks for the additional section. I added some more nitpicks, mind taking a look again?

**Comment by kmaehashi** [comment]
Thanks, fixed accordingly.

File: docs/source/guides/variables.rst [hunk 7]
```diff
@@ -70,6 +76,46 @@ This is done simply by setting the :attr:`~Variable.grad` attribute of the outpu
...
+
+
+Higher-Order Derivatives
+------------------------
+
+:class:`~chainer.Variable` also supports higher-order derivatives (a.k.a. double backpropagation).
+
+Let's see a simple example.
+First calculate the first-order derivative.
+Note that ``enable_double_backprop=True`` is pass to ``y.backward()``.
```
Comment (hvy) [inline docs/source/guides/variables.rst]: Ops, typo?

Comment (kmaehashi) [reply docs/source/guides/variables.rst]: 🙇

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
@niboshi Could you check and approve the review?

File: docs/source/guides/variables.rst [hunk 8]
```diff
@@ -1,31 +1,37 @@
...
 
-   >>> y.data
+   >>> y.array
    array([16.], dtype=float32)
 
+.. note::
+
+   :attr:`~chainer.Variable` has two attributes to represent the underlying array: :attr:`~chainer.Variable.array` and :attr:`~chainer.Variable.data`.
+   There is no difference between the two; both refer to exactly the same object.
+   However it is not recommended to use ``.data`` because it because it might be confused with :attr:`numpy.ndarray.data` attribute.
```
Comment (niboshi) [inline docs/source/guides/variables.rst]: Erroneous repetition of "becuase it".

Comment (kmaehashi) [reply docs/source/guides/variables.rst]: Oops, fixed.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
LGTM, the CI failure is not related to this PR.

**Pull request merged**